finality-aleph/src/sync/handler.rs

```diff
@@ -95,130 +101,57 @@ impl<I: PeerId, J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finaliz
+                    .map_err(Error::Finalizer)?;
+                number += 1;
             }
-            number += 1;
-        }
-        Ok(NetworkData::Justifications(result, self.state()?))
-    }
-
-    fn top_understandable_for(
-        &self,
-        id: BlockIdFor<J>,
-    ) -> Result<NetworkData<J>, Error<J, CS, V, F>> {
-        use Error::*;
-        let block_session = session_id_from_block_num(id.number(), self.period);
-        let understandable_session = SessionId(block_session.0 + 1);
-        let last_understandable_block_number =
-            last_block_of_session(understandable_session, self.period);
-        match self
-            .chain_status
-            .finalized_at(last_understandable_block_number)
-            .map_err(ChainStatus)?
-        {
-            Some(justification) => Ok(NetworkData::Justifications(
-                vec![justification.into_unverified()],
-                self.state()?,
-            )),
-            None => {
-                let justification = self.chain_status.top_finalized().map_err(ChainStatus)?;
-                match justification.header().id().number() <= last_understandable_block_number {
-                    true => Ok(NetworkData::Justifications(
-                        vec![justification.into_unverified()],
-                        self.state()?,
-                    )),
-                    false => Err(Error::MissingJustification),
+            number =
+                last_block_of_session(session_id_from_block_num(number, self.period), self.period);
+            match self.forest.try_finalize(&number) {
+                Some(justification) => {
+                    self.finalizer
+                        .finalize(justification)
+                        .map_err(Error::Finalizer)?;
+                    number += 1;
                 }
-            }
+                None => return Ok(()),
+            };
         }
     }
 
-    fn try_finalize(&mut self) -> Result<(), Error<J, CS, V, F>> {
-        while let Some(justification) = self.forest.try_finalize() {
-            self.finalizer
-                .finalize(justification)
-                .map_err(Error::Finalizer)?;
-        }
-        Ok(())
-    }
-
     fn handle_verified_justification(
         &mut self,
         justification: J,
         peer: I,
     ) -> Result<SyncActions<J>, Error<J, CS, V, F>> {
-        use JustificationAddResult::*;
         let id = justification.header().id();
-        match self
+        let action = match self
             .forest
             .update_justification(justification, Some(peer))?
         {
-            Noop => Ok(SyncActions::noop()),
-            Required => Ok(SyncActions::task(id)),
-            Finalizable => {
-                self.try_finalize()?;
-                Ok(SyncActions::noop())
-            }
-        }
+            true => Ok(SyncActions::task(id)),
+            false => Ok(SyncActions::noop()),
+        };
+        self.try_finalize()?;
```

Comment:
I feel this is too specific, the `Forest` logic changes so often that this might come back to bite us. Let's just keep the rule that you call `try_finalize` after every block of other calls, as this is cheap to do.